go/tools/gazelle/rules/resolve_external.go

```diff
@@ -70,22 +70,55 @@ func (e externalResolver) resolve(importpath, dir string) (label, error) {
 	}, nil
 }
 
-// knownImports are paths which are not static in the vcs package,
-// to allow load balancing between actual repos,
-// but for our case we only need to break the importpath in a known fashion.
-var knownImports = []string{"golang.org/x/", "google.golang.org/", "cloud.google.com/"}
+func findCachedRepoRoot(importpath string) string {
+	// subpaths contains slices of importpath with components removed. For
+	// example:
+	//   golang.org/x/tools/go/vcs
+	//   golang.org/x/tools/go
+	//   golang.org/x/tools
+	subpaths := []string{importpath}
 
-// specialCases looks for matches in knownImports to avoid making a network call.
-func specialCases(importpath string) string {
-	for _, known := range knownImports {
-		if !strings.HasPrefix(importpath, known) {
-			continue
+	for {
+		if n, ok := repoRootCache[importpath]; ok {
+			if n >= len(subpaths) {
+				// The import path is shorter than expected. Treat as a miss.
+				return ""
+			}
+			// Cache hit.
+			return subpaths[len(subpaths)-n-1]
```

Comment:
Sorry about that, I know it's confusing. I thought about doing something with splitting and joining strings, but I thought it might be just as confusing and less efficient.

I've updated the comment.